collector/perf_linux.go

```diff
@@ -0,0 +1,549 @@
+// Copyright 2019 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"fmt"
+	"runtime"
+	"sync"
+
+	perf "github.com/hodgesds/perf-utils"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+)
+
+const (
+	perfSubsystem = "perf"
+)
+
+var (
+	perfHwProfilers    = map[int]perf.HardwareProfiler{}
+	perfSwProfilers    = map[int]perf.SoftwareProfiler{}
+	perfCacheProfilers = map[int]perf.CacheProfiler{}
+	perfInitializeOnce sync.Once
+)
+
+func init() {
+	registerCollector(perfSubsystem, defaultDisabled, NewPerfCollector)
+}
+
+// perfCollector is a Collected that uses the perf subsystem to collect
+// metrics. It uses perf_event_open an ioctls for profiling.
+type perfCollector struct{}
```

Comment:
Why this additional struct? Can you make it member of perfCollector? Then you also don't need sync.Once.

Reply:
The profilers are meant to be long lived, within the library it opens up file descriptors for the individual events being profiled. Since `NewPerfCollector` is called every collection cycle the newly allocated struct either needs to use an embedded global variable within it or just use some other global variable (which is the current approach). I'm a fan of the latter approach because it doesn't require copying the maps around. If you see in the comments in this PR I ran into that issue while working on this. I've added some additional notes in the code for other to better understand.

Reply:
Now it's fine. We just don't need to have two structs. No need to copying around maps either.